sentry/src/main/java/io/sentry/clientreport/AtomicClientReportStorage.java

```diff
@@ -0,0 +1,76 @@
+package io.sentry.clientreport;
+
+import io.sentry.ILogger;
+import io.sentry.SentryLevel;
+import io.sentry.SentryOptions;
+import io.sentry.transport.DataCategory;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+final class AtomicClientReportStorage implements ClientReportStorage {
```

Comment:
The idea here is to pre-initialize all possible combinations of `DiscardReason` and `DataCategory` that can ocurr with an `AtomicLong` of value `0`. Then for all future count increases the access is synchronized only for the specific combination using `AtomicLong` instead of across all operations (like `LockingClientReportRecorder` does).

Reply:
Let's not forget to filter out zeroed reasons before sending to Sentry tho.

Reply:
This happens further down:

```
if (quantity > 0) {
...
```